### ISSUE TITLE
Error in dice loss computation

### ISSUE THREAD

**Issue by BernhardSt** [milesial/Pytorch-UNet#24]
Hi, 

first of all thanks a lot for open sourcing this repo.

There is an error in the dice loss computation in dice_loss.py. 
Your implementation :
`self.inter = torch.dot(input.view(-1), target.view(-1)) + 0.0001`
`self.union = torch.sum(input) + torch.sum(target) + 0.0001`
` t = 2 * self.inter.float() / self.union.float()`

The dice loss is defined by: `2 * intersection / union`, i.e., `(2 * dot(A,B)) / (dot(A,A) + dot(B,B)) `. A small number⁠ ⁠⁠ɛ⁠ = 0.0001 is added to the intersection and union term in the implementation to avoid division by zero. 
While the epsilon is important, it is wrongly placed. The correct version should be to first compute the intersection including the times two factor and then adding the epsilon: `(2 * dot(A,B) + ɛ) / (dot(A,A) + dot(B,B)) + ɛ⁠)`.  

Usually, this is just a minor flaw that incorrectly computes the dice by the amount of 1*epsilon because the epsilon in the nominator is multiplied by 2. However, for an empty label and prediction map, e.g., in case of correctly deciding that nothing should be segmented in the image, the current form evaluates to 2, which is obviously not defined for the dice score. 

To compare, for  `dot(A,B) = 0`:
Current implementation:
`2 * (0 + ɛ⁠)/(0 + 0 + ɛ⁠)  = 2 * ɛ/ɛ   = 2 * 1 `
Correct:
`((2 * 0) + ɛ⁠)/((0 + 0) + ɛ⁠)  = ɛ/ɛ = 1`

Best Regards and Thanks again!

**Comment by milesial** [comment]
Hi, thanks for the detailed explaination ! I fixed it : https://github.com/milesial/Pytorch-UNet/commit/1e8093de0398364cc65f3c7722b61a0f560df104
